# Introduction to Pandas

*Author: [Douglas Strodtman](http://linkedin.com/in/dstrodtman/)*

In this notebook, we'll cover the basic functionality of Pandas, which is one of the most widely used Python packages for data exploration, cleaning, and manipulation. The official docs have a [10 Minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html) lesson; we'll take a bit longer and hopefully generalize concepts to have a deeper understanding.

If you have robust knowledge working with R, SQL, SAS, or Stata, I recommend that you read through [the corresponding comparison notes provided by the maintainers of Pandas](http://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/).

## A Note on the Data

The city of Los Angeles has adopted one of the most progressive approaches to open data in the nation under the leadership of Mayor Garcetti.

The [Los Angeles Open Data portal](https://data.lacity.org/) is an amazing resource for accessing city-specific data. It has a number of built-in tools to let you explore data in the browser prior to downloading, and will let you export data in many popular formats.

Many online tutorials will have you work with canonical datasets that are well-documented. While these are fine, I encourage you to always work with real data when possible, and to apply your skills to questions that you'd like to answer.

Today, we'll be working with [City Budget and Expenditures](https://controllerdata.lacity.org/Budget/City-Budget-and-Expenditures/uyzw-yi8n) data. The version that has been included with this lesson was updated on March 8, 2019.

The data is presented as a line item budget, showing aggregates for accounts by department and fund. We can consider each row in our data to be a line item.

## Lesson Overview

1. Module import
1. Data import
1. Preview Data
    - head
    - tail
    - sample
1. DataFrames and Series
    - What are they?
    - Data selection
        - One column
        - Multiple columns
        - One row
        - Multiple rows
        - Singular values
1. Metadata exploration
    - Data size
    - Data types
    - Columns
    - Index
1. Methods and Attributes
1. Categorical data
    - Unique values
    - Value counts
    - String operations
1. Numeric data
    - Summary statistics
    - Basic math operations
1. Sorting


## Module Import

`pandas` is always imported as `pd`. It's good to import all modules that you might need in a notebook in the first code block.

In [1]:
import pandas as pd

Pandas is in active open source development. While the maintainers do a good job trying to make sure that newer version are backwards compatible, occassionally there will be breaking changes.

Checking the `__version__` of an imported module will let us know if we're all in the same development environment. The newest version is 0.24.2 (at the time of this writing, April 2019). It is unlikely that you will see significant changes in performance as long as your version is > 0.20.0.

In [2]:
pd.__version__

'0.23.4'

## Data Import

Pandas provides support for many different data file types. Read the docs [here](https://pandas.pydata.org/pandas-docs/stable/reference/io.html) if you need help loading a file.

Provided with this lesson are a number of data files. We'll begin our exploration by loading the file `2018_budget.csv`.

When you read data into Pandas, it will be stored into a DataFrame (we'll explore what this means shortly).

In order to explore and manipulate our DataFrame, we'll assign it to a variable.

### A Note on Naming Conventions

Python has many naming conventions, and some of these vary slightly based on area of practice.

Generally speaking, it's best practice to use [snake case](https://en.wikipedia.org/wiki/Snake_case) for all variable names, and to provide your variables with informative human-readable names. Initial capital letters are usually reserved for `Class`es and `Type`s and written in [camel case](https://en.wikipedia.org/wiki/Camel_case). Variable names in all caps are usually reserved for global variables.

| casing | use case | example |
| --- | --- | --- |
| `snake_case` | most variables and functions | `some_variable`, `some_function` |
| `CamelCase` | classes | `DataFrame`, `Series` |
| `UPPERCASE` | global variables | `USER_CREDENTIALS`, `IP_ADDRESS` |

Until you become a more advanced Python user, you will probably want to exclusively use `snake_case`, as you are unlikely to be defining classes or global variables.

### A Specific Note on `df`

As you are searching for solutions to Pandas related issues and going through online tutorials, you will find a proliferation of the variable name `df`.

`df` is shorthand for DataFrame, and you can readily adapt code you find online to your use case by changing this variable name to make your user-defined variable name.

> "Should I use `df` to name _my_ DataFrame?" - Literally Everyone

_**Maybe**_.

My rule of thumb is that if I know I'm only going to be loading in a single DataFrame, I'll stick with `df`. If I'm going to have 2 of more DataFrames in a single notebook, then I always give each a descriptive, human-readable name.

**Note**: Please don't ever just append numbers to differentiate DataFrames. This creates a huge mental burden on you as a programmer and your audience to keep track of what data is in each DataFrame.

### Load Data as `df` with `read_csv`

Here, we'll use the standard `df` to read our CSV with default options.

Pandas is great about having generally good default options that do what you probably want, but if your data ever loads in incorrectly, you can probably set an option or two to fix this by [consulting the docs](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html#pandas.read_csv).

In [3]:
df = pd.read_csv('../data/2018_budget.csv')

## Preview Data

Pandas is meant for computational evaluation and manipulation of your data. This means that you'll never be manually poring over every value in your data. Our data is going to be big, and visual inspection will be a waste of our time.

Similarly, because our data is big, we don't want to ever print out all rows. Instead, we'll check that our manipulations are working correctly by printing out small previews.

### `.head`

By default, this returns the top 5 rows. Pass a scalar value as an argument to get that number of rows returned.

In [4]:
df.head()

budget_fiscal_year department_name                      fund_name  \
0                2018           AGING  GENERAL FUND (GENERAL BUDGET)   
1                2018           AGING  TITLE VII OLDER AMERICANS ACT   
2                2018           AGING  SENIOR HUMAN SERVICES PROGRAM   
3                2018           AGING  AREA PLAN FOR THE AGING TIT 7   
4                2018           AGING  GENERAL FUND (GENERAL BUDGET)   

                       account_name  adopted_budget_amount  \
0              CONTRACTUAL SERVICES              2222382.0   
1           OMBUDSMAN VII A PROGRAM                    0.0   
2           EVIDENCE BASED PROGRAMS                    0.0   
3  HOME DELIVERED MEALS FOR SENIORS                    0.0   
4                  OVERTIME GENERAL                 3900.0   

   total_expenditures  budget_change_amount  budget_transfer_in_amount  \
0          1608157.04                9000.0                        0.0   
1            87876.00               87876.0                        0.0   
2           292338.00              303447.0                        0.0   
3          2419162.00             2543845.0                        0.0   
4            15943.36                   0.0                    13300.0   

   budget_transfer_out_amount  total_budget  encumbrance_amount  \
0                    453500.0     1777882.0             93331.0   
1                         0.0       87876.0                 0.0   
2                         0.0      303447.0             11109.0   
3                         0.0     2543845.0             36122.0   
4                         0.0       17200.0                 0.0   

   pre_encumbrance_amount  budget_uncommitted_amount     account_group_name  \
0                     0.0                   76393.96               EXPENSES   
1                     0.0                       0.00                    NaN   
2                     0.0                       0.00                    NaN   
3                     0.0                   88561.00                    NaN   
4                     0.0                    1256.64  SALARIES AND BENEFITS   

  fund account  department  
0  100  003040           2  
1  564  02PB01           2  
2  42J  02R340           2  
3  395  02PQ04           2  
4  100  001090           2

In [5]:
df.head(3)

budget_fiscal_year department_name                      fund_name  \
0                2018           AGING  GENERAL FUND (GENERAL BUDGET)   
1                2018           AGING  TITLE VII OLDER AMERICANS ACT   
2                2018           AGING  SENIOR HUMAN SERVICES PROGRAM   

              account_name  adopted_budget_amount  total_expenditures  \
0     CONTRACTUAL SERVICES              2222382.0          1608157.04   
1  OMBUDSMAN VII A PROGRAM                    0.0            87876.00   
2  EVIDENCE BASED PROGRAMS                    0.0           292338.00   

   budget_change_amount  budget_transfer_in_amount  \
0                9000.0                        0.0   
1               87876.0                        0.0   
2              303447.0                        0.0   

   budget_transfer_out_amount  total_budget  encumbrance_amount  \
0                    453500.0     1777882.0             93331.0   
1                         0.0       87876.0                 0.0   
2                         0.0      303447.0             11109.0   

   pre_encumbrance_amount  budget_uncommitted_amount account_group_name fund  \
0                     0.0                   76393.96           EXPENSES  100   
1                     0.0                       0.00                NaN  564   
2                     0.0                       0.00                NaN  42J   

  account  department  
0  003040           2  
1  02PB01           2  
2  02R340           2

### `.tail`

The same as `.head` but for the end of our DataFrame.

In [6]:
df.tail(3)

budget_fiscal_year department_name                      fund_name  \
3650                2018             ZOO      ZOO ENTERPRISE TRUST FUND   
3651                2018             ZOO  GENERAL FUND (GENERAL BUDGET)   
3652                2018             ZOO  GENERAL FUND (GENERAL BUDGET)   

                       account_name  adopted_budget_amount  \
3650        ZOO WASTEWATER FACILITY                    0.0   
3651               OVERTIME GENERAL               135164.0   
3652  VETERINARY SUPPLIES & EXPENSE               400000.0   

      total_expenditures  budget_change_amount  budget_transfer_in_amount  \
3650           242592.00             246529.75                        0.0   
3651           164024.80                  0.00                    35000.0   
3652           309926.01                  0.00                        0.0   

      budget_transfer_out_amount  total_budget  encumbrance_amount  \
3650                         0.0     246529.75                0.00   
3651                         0.0     170164.00                0.00   
3652                         0.0     400000.00            37915.63   

      pre_encumbrance_amount  budget_uncommitted_amount  \
3650                     0.0                    3937.75   
3651                     0.0                    6139.20   
3652                     0.0                   52158.36   

         account_group_name fund account  department  
3650                    NaN  40E  87P914          87  
3651  SALARIES AND BENEFITS  100  001090          87  
3652               EXPENSES  100  004570          87

### `.sample`

Provides a random sample of the rows in our data.

In [7]:
df.sample(5)

budget_fiscal_year              department_name  \
513                 2018                   CITY CLERK   
495                 2018                   CITY CLERK   
1180                2018                         FIRE   
206                 2018  CITY ADMINISTRATIVE OFFICER   
862                 2018             CULTURAL AFFAIRS   

                           fund_name  \
513   SUNSHINE CYN COMMUNI AMENITIES   
495        NEIGHBORHOOD COUNCIL FUND   
1180  MICLA LEASE REV COM PAPER NTES   
206   PROPOSITION K MAINTENANCE FUND   
862   ARTS DEVELOPMENT FEE TRUST FND   

                                           account_name  \
513                               BUREAU OF SANITATIONS   
495                WESTLAKE SOUTH  NEIGHBORHOOD COUNCIL   
1180                   DIESEL EXHAUST CAPTURING SYSTEMS   
206   MAINT: CALIFORNIA HOSPITAL MEDICAL CENTER (CHM...   
862                  3400-3404 W SUNSET 90026 PMT 02786   

      adopted_budget_amount  total_expenditures  budget_change_amount  \
513                     0.0            25862.65              25862.65   
495                     0.0            42000.00              42000.00   
1180                    0.0           962461.87            2390000.00   
206                     0.0             4458.96              13416.00   
862                     0.0                 NaN               1268.69   

      budget_transfer_in_amount  budget_transfer_out_amount  total_budget  \
513                         0.0                         0.0      25862.65   
495                         0.0                         0.0      42000.00   
1180                        0.0                         0.0    2390000.00   
206                         0.0                         0.0      13416.00   
862                         0.0                         0.0       1268.69   

      encumbrance_amount  pre_encumbrance_amount  budget_uncommitted_amount  \
513                 0.00                     0.0                       0.00   
495                 0.00                     0.0                       0.00   
1180             3762.42                     0.0                 1423775.71   
206              8957.04                     0.0                       0.00   
862                 0.00                     0.0                    1268.69   

     account_group_name fund account  department  
513                 NaN  699  14P182          14  
495                 NaN  59B  141289          14  
1180                NaN  298  38PDEC          38  
206             SPECIAL  43L  10PMAF          10  
862                 NaN  516  30PA28          30

### Note on Data Previews

While we'll use `head` extensively throughout this demo, it's generally good practice to validate your operations by sampling from throughout your data, especially as you're learning (as you're especially prone to make mistakes that might not be apparent in just the first 5 rows).

In particular, manual inspection of the `tail` of your data will often quickly reveal if there was an error while parsing/loading your data, as generally these errors will cause misalignment that propagates throughout your data and will be easily visible in your last few rows.

## `DataFrames` and `Series`

The two main data structures you'll need to understand when working with Pandas are `DataFrames` and `Series`.

In brief, Pandas provides tabular (table-based) access to data, porting much of the functionality of Excel and SQL into the Python environment. As such, we'll be talking a lot about rows, columns, and cells. 

At the same time, it is implemented to be optimized for vectorized operations, meaning that we can also think of a DataFrame as a matrix, with our Series representing vectors, and our cells being scalars. **If you don't speak linear algebra, don't worry.** If you just use Pandas built-in functions, you should see the performance benefits without ever thinking about what a vector is.

Here's a table to help you organize these concepts:

| Pandas structure | Common name | Linalg complement | Dimensions |
| --- | --- | --- | --- |
| DataFrame | table, sheet | matrix | 2-d (rows by columns) |
| Series | column, row | vector, array | 1-d (rows and columns have length only) |
| values | cell | scalar | 0-d (these are single values) |


If you're inclined, you can get a deep understanding of these by [reading the docs](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html).

For our purposes, we'll make a few simple (but important) distinctions.

### `Series`

We can think of Series as the rows and columns of a DataFrame. When we access a series, the returned preview won't be pretty printed.

#### Columns

We can access Series in a number of ways. Most commonly, we'll use a single square bracket notation.

**NOTE**: We'll use `head` on all of our column operations to only show the first 5 rows.

In [8]:
df['fund_name'].head()

0    GENERAL FUND (GENERAL BUDGET)
1    TITLE VII OLDER AMERICANS ACT
2    SENIOR HUMAN SERVICES PROGRAM
3    AREA PLAN FOR THE AGING TIT 7
4    GENERAL FUND (GENERAL BUDGET)
Name: fund_name, dtype: object

As long as our column names are only letters, numbers, and underscores, we can also use a dot notation to access Series.

In [9]:
df.total_budget.head()

0    1777882.0
1      87876.0
2     303447.0
3    2543845.0
4      17200.0
Name: total_budget, dtype: float64

We can also use `.loc` notation to access a Series.

In [10]:
df.loc[:, 'account_name'].head()

0                CONTRACTUAL SERVICES
1             OMBUDSMAN VII A PROGRAM
2             EVIDENCE BASED PROGRAMS
3    HOME DELIVERED MEALS FOR SENIORS
4                    OVERTIME GENERAL
Name: account_name, dtype: object

The `:` here indicates all rows. 

We can also select a specific range of rows.

In [11]:
df.loc[10:15, 'account_name']

10                                TRAVEL
11                                 AGING
12          CONGREGATE MEALS FOR SENIORS
13      OTO NSIP CONGREGATE MEALS III C1
14           FINANCIAL ALIGNMENT PROGRAM
15    ELDER ABUSE PREV PROGR FOR SENIORS
Name: account_name, dtype: object

We can also provide a list of specific rows we want to include.

In [12]:
df.loc[[10,12,27,45,500], 'account_name']

10                             TRAVEL
12       CONGREGATE MEALS FOR SENIORS
27                      HICAP 3 MONTH
45                   SALARIES GENERAL
500    WESTWOOD  NEIGHBORHOOD COUNCIL
Name: account_name, dtype: object

#### Rows

To return an entire row, just use `.loc` with the index label for that row.

In [13]:
df.loc[0]

budget_fiscal_year                                     2018
department_name                                       AGING
fund_name                     GENERAL FUND (GENERAL BUDGET)
account_name                           CONTRACTUAL SERVICES
adopted_budget_amount                           2.22238e+06
total_expenditures                              1.60816e+06
budget_change_amount                                   9000
budget_transfer_in_amount                                 0
budget_transfer_out_amount                           453500
total_budget                                    1.77788e+06
encumbrance_amount                                    93331
pre_encumbrance_amount                                    0
budget_uncommitted_amount                             76394
account_group_name                                 EXPENSES
fund                                                    100
account                                              003040
department                              

Similar to how we specified rows for our column selections, we can specify columns for our row selections.

In [14]:
df.loc[0, ['fund_name', 'account_name', 'total_budget']]

fund_name       GENERAL FUND (GENERAL BUDGET)
account_name             CONTRACTUAL SERVICES
total_budget                      1.77788e+06
Name: 0, dtype: object

We can use the same `:` notation to select a range of our columns (note that the columns returned are dependent upon the order of columns in your DataFrame).

In [15]:
df.loc[0, 'fund_name':'total_budget']

fund_name                     GENERAL FUND (GENERAL BUDGET)
account_name                           CONTRACTUAL SERVICES
adopted_budget_amount                           2.22238e+06
total_expenditures                              1.60816e+06
budget_change_amount                                   9000
budget_transfer_in_amount                                 0
budget_transfer_out_amount                           453500
total_budget                                    1.77788e+06
Name: 0, dtype: object

#### Quick Note on `.iloc`

Pandas also provides an `.iloc` selection method. This uses the numeric index of rows and columns to handle selection. While this can be useful in some applications, most times in Pandas you will have informative column and index labels that you would rather select on.

When our index is a serial integer range (as here), `.loc` and `.iloc` have similar behavior when selecting rows. `.iloc` will not allow you to pass column names, however.

In [16]:
df.iloc[0]

budget_fiscal_year                                     2018
department_name                                       AGING
fund_name                     GENERAL FUND (GENERAL BUDGET)
account_name                           CONTRACTUAL SERVICES
adopted_budget_amount                           2.22238e+06
total_expenditures                              1.60816e+06
budget_change_amount                                   9000
budget_transfer_in_amount                                 0
budget_transfer_out_amount                           453500
total_budget                                    1.77788e+06
encumbrance_amount                                    93331
pre_encumbrance_amount                                    0
budget_uncommitted_amount                             76394
account_group_name                                 EXPENSES
fund                                                    100
account                                              003040
department                              

In [17]:
df.iloc[0:5, 5]

0    1608157.04
1      87876.00
2     292338.00
3    2419162.00
4      15943.36
Name: total_expenditures, dtype: float64

In [18]:
df.iloc[0, 2:7]

fund_name                GENERAL FUND (GENERAL BUDGET)
account_name                      CONTRACTUAL SERVICES
adopted_budget_amount                      2.22238e+06
total_expenditures                         1.60816e+06
budget_change_amount                              9000
Name: 0, dtype: object

As we proceed forward and insert, sort, and remove rows and columns, it will become apparent why using specific labels makes for clearer, more robust code for selection and manipulation.

### `DataFrames`

DataFrames are tables, essentially, or a collection of rows and columns. When we access a DataFrame, the returned preview will be pretty printed.

**NOTE**: We'll use `head` on all of our DataFrame operations to only show the first 5 rows.

To show an entire DataFrame, just type in the name and evaluate the code.

In [19]:
df.head()

budget_fiscal_year department_name                      fund_name  \
0                2018           AGING  GENERAL FUND (GENERAL BUDGET)   
1                2018           AGING  TITLE VII OLDER AMERICANS ACT   
2                2018           AGING  SENIOR HUMAN SERVICES PROGRAM   
3                2018           AGING  AREA PLAN FOR THE AGING TIT 7   
4                2018           AGING  GENERAL FUND (GENERAL BUDGET)   

                       account_name  adopted_budget_amount  \
0              CONTRACTUAL SERVICES              2222382.0   
1           OMBUDSMAN VII A PROGRAM                    0.0   
2           EVIDENCE BASED PROGRAMS                    0.0   
3  HOME DELIVERED MEALS FOR SENIORS                    0.0   
4                  OVERTIME GENERAL                 3900.0   

   total_expenditures  budget_change_amount  budget_transfer_in_amount  \
0          1608157.04                9000.0                        0.0   
1            87876.00               87876.0                        0.0   
2           292338.00              303447.0                        0.0   
3          2419162.00             2543845.0                        0.0   
4            15943.36                   0.0                    13300.0   

   budget_transfer_out_amount  total_budget  encumbrance_amount  \
0                    453500.0     1777882.0             93331.0   
1                         0.0       87876.0                 0.0   
2                         0.0      303447.0             11109.0   
3                         0.0     2543845.0             36122.0   
4                         0.0       17200.0                 0.0   

   pre_encumbrance_amount  budget_uncommitted_amount     account_group_name  \
0                     0.0                   76393.96               EXPENSES   
1                     0.0                       0.00                    NaN   
2                     0.0                       0.00                    NaN   
3                     0.0                   88561.00                    NaN   
4                     0.0                    1256.64  SALARIES AND BENEFITS   

  fund account  department  
0  100  003040           2  
1  564  02PB01           2  
2  42J  02R340           2  
3  395  02PQ04           2  
4  100  001090           2

We can use double square bracket notation to select a subset of columns.

In [20]:
df[['department_name', 'fund_name', 'account_name']].head()

department_name                      fund_name  \
0           AGING  GENERAL FUND (GENERAL BUDGET)   
1           AGING  TITLE VII OLDER AMERICANS ACT   
2           AGING  SENIOR HUMAN SERVICES PROGRAM   
3           AGING  AREA PLAN FOR THE AGING TIT 7   
4           AGING  GENERAL FUND (GENERAL BUDGET)   

                       account_name  
0              CONTRACTUAL SERVICES  
1           OMBUDSMAN VII A PROGRAM  
2           EVIDENCE BASED PROGRAMS  
3  HOME DELIVERED MEALS FOR SENIORS  
4                  OVERTIME GENERAL

We could have also selected a single column in this way, still returning a DataFrame.

**NOTE**: While here we look at a single column, we still have a DataFrame rather than a Series. Certain Pandas operations require Series and will error out if we try to pass a DataFrame instead.

In [21]:
df[['fund_name']].head()

fund_name
0  GENERAL FUND (GENERAL BUDGET)
1  TITLE VII OLDER AMERICANS ACT
2  SENIOR HUMAN SERVICES PROGRAM
3  AREA PLAN FOR THE AGING TIT 7
4  GENERAL FUND (GENERAL BUDGET)

We can nest brackets within our `.loc` notation to return a DataFrame instead of a Series.

In [22]:
df.loc[:, ['department_name','total_expenditures']].head()

department_name  total_expenditures
0           AGING          1608157.04
1           AGING            87876.00
2           AGING           292338.00
3           AGING          2419162.00
4           AGING            15943.36

This is especially useful when we're interested in looking at multiple columns for a range of rows.

In [23]:
df.loc[0:10, ['department_name','total_expenditures']]

department_name  total_expenditures
0            AGING          1608157.04
1            AGING            87876.00
2            AGING           292338.00
3            AGING          2419162.00
4            AGING            15943.36
5            AGING           256193.00
6            AGING           218106.00
7            AGING          2740986.00
8            AGING          4393789.00
9            AGING          1444876.40
10           AGING             5347.79

Or when we're interested in selecting specific rows with multiple columns.

In [24]:
df.loc[[0,10,20,30,40], ['fund_name', 'adopted_budget_amount']]

fund_name  adopted_budget_amount
0   GENERAL FUND (GENERAL BUDGET)              2222382.0
10  GENERAL FUND (GENERAL BUDGET)                 8650.0
20  AREA PLAN FOR THE AGING TIT 7                    0.0
30  AREA PLAN FOR THE AGING TIT 7                    0.0
40  GENERAL FUND (GENERAL BUDGET)               222431.0

If we nest brackets on just our rows, we'll again get a DataFrame rather than a Series.

In [25]:
df.loc[[0]]

budget_fiscal_year department_name                      fund_name  \
0                2018           AGING  GENERAL FUND (GENERAL BUDGET)   

           account_name  adopted_budget_amount  total_expenditures  \
0  CONTRACTUAL SERVICES              2222382.0          1608157.04   

   budget_change_amount  budget_transfer_in_amount  \
0                9000.0                        0.0   

   budget_transfer_out_amount  total_budget  encumbrance_amount  \
0                    453500.0     1777882.0             93331.0   

   pre_encumbrance_amount  budget_uncommitted_amount account_group_name fund  \
0                     0.0                   76393.96           EXPENSES  100   

  account  department  
0  003040           2

This can be used to select one or many rows, with one or many columns.

In [26]:
df.loc[[50,51,52,53,54,55]]

budget_fiscal_year department_name                       fund_name  \
50                2018        AIRPORTS    AIRPORT REVENUE FUND-ONTARIO   
51                2018        AIRPORTS                 AIRPORT REVENUE   
52                2018        AIRPORTS                 AIRPORT REVENUE   
53                2018        AIRPORTS  DEA FEDERAL FORFEIT PROP-LAWA/   
54                2018        AIRPORTS    AIRPORT REVENUE FUND-ONTARIO   
55                2018        AIRPORTS  PASSENGER FACILITY CHARGE -LAX   

            account_name  adopted_budget_amount  total_expenditures  \
50      SALARIES GENERAL                    0.0        5.909209e+06   
51  HIRING HALL-OVERTIME                    0.0        8.864508e+04   
52    OTHER EXPENDITURES                    0.0        1.943900e+09   
53    OTHER EXPENDITURES                    0.0        1.821408e+05   
54    OTHER EXPENDITURES                    0.0        8.205629e+06   
55    OTHER EXPENDITURES                    0.0        2.737955e+08   

    budget_change_amount  budget_transfer_in_amount  \
50                   0.0                        0.0   
51                   0.0                        0.0   
52                   0.0                        0.0   
53                   0.0                        0.0   
54                   0.0                        0.0   
55                   0.0                        0.0   

    budget_transfer_out_amount  total_budget  encumbrance_amount  \
50                         0.0           0.0                 0.0   
51                         0.0           0.0                 0.0   
52                         0.0           0.0                 0.0   
53                         0.0           0.0                 0.0   
54                         0.0           0.0                 0.0   
55                         0.0           0.0                 0.0   

    pre_encumbrance_amount  budget_uncommitted_amount account_group_name fund  \
50                     0.0              -5.909209e+06                NaN  723   
51                     0.0              -8.864508e+04                NaN  700   
52                     0.0              -1.943900e+09                NaN  700   
53                     0.0              -1.821408e+05                NaN  74L   
54                     0.0              -8.205629e+06                NaN  723   
55                     0.0              -2.737955e+08                NaN  71R   

   account  department  
50  041010           4  
51  041190           4  
52  041000           4  
53  041000           4  
54  041000           4  
55  041000           4

#### A Note on Transpose

DataFrames contain a transpose attribute, `.T`. This doesn't change the underlying data, but just changes the orientation of the returned preview, switching the location or rows with columns.

In [27]:
df.T.head()

0     \
budget_fiscal_year                              2018   
department_name                                AGING   
fund_name              GENERAL FUND (GENERAL BUDGET)   
account_name                    CONTRACTUAL SERVICES   
adopted_budget_amount                    2.22238e+06   

                                                1     \
budget_fiscal_year                              2018   
department_name                                AGING   
fund_name              TITLE VII OLDER AMERICANS ACT   
account_name                 OMBUDSMAN VII A PROGRAM   
adopted_budget_amount                              0   

                                                2     \
budget_fiscal_year                              2018   
department_name                                AGING   
fund_name              SENIOR HUMAN SERVICES PROGRAM   
account_name                 EVIDENCE BASED PROGRAMS   
adopted_budget_amount                              0   

                                                   3     \
budget_fiscal_year                                 2018   
department_name                                   AGING   
fund_name                 AREA PLAN FOR THE AGING TIT 7   
account_name           HOME DELIVERED MEALS FOR SENIORS   
adopted_budget_amount                                 0   

                                                4     \
budget_fiscal_year                              2018   
department_name                                AGING   
fund_name              GENERAL FUND (GENERAL BUDGET)   
account_name                        OVERTIME GENERAL   
adopted_budget_amount                           3900   

                                                             5     \
budget_fiscal_year                                           2018   
department_name                                             AGING   
fund_name                          OMBUDSMAN INITIATIVE PROGRAM F   
account_name           STATE HEALTH FACILITIES CITATION PENALTIES   
adopted_budget_amount                                           0   

                                                6     \
budget_fiscal_year                              2018   
department_name                                AGING   
fund_name              AREA PLAN FOR THE AGING TIT 7   
account_name                 PREVENTIVE HEALTH III D   
adopted_budget_amount                              0   

                                                7     \
budget_fiscal_year                              2018   
department_name                                AGING   
fund_name              AREA PLAN FOR THE AGING TIT 7   
account_name             HOME DELIVERED MEALS III C2   
adopted_budget_amount                              0   

                                                8     \
budget_fiscal_year                              2018   
department_name                                AGING   
fund_name              AREA PLAN FOR THE AGING TIT 7   
account_name                 CONGREGATE MEALS III C1   
adopted_budget_amount                              0   

                                               9     \
budget_fiscal_year                             2018   
department_name                               AGING   
fund_name              OTHER PROGRAMS FOR THE AGING   
account_name                         ENROLLEE WAGES   
adopted_budget_amount                             0   

                                   ...                \
budget_fiscal_year                 ...                 
department_name                    ...                 
fund_name                          ...                 
account_name                       ...                 
adopted_budget_amount              ...                 

                                                3643  \
budget_fiscal_year                              2018   
department_name                                  ZOO   
fund_name              GENERAL FUND (GENERAL BUDGET)   
account_name               

We can apply this to any of our above DataFrame access methods it we prefer to our data in this transposed configuration.

In [28]:
df.loc[[0,10,20,30,40], ['fund_name', 'adopted_budget_amount']].T

0   \
fund_name              GENERAL FUND (GENERAL BUDGET)   
adopted_budget_amount                    2.22238e+06   

                                                  10  \
fund_name              GENERAL FUND (GENERAL BUDGET)   
adopted_budget_amount                           8650   

                                                  20  \
fund_name              AREA PLAN FOR THE AGING TIT 7   
adopted_budget_amount                              0   

                                                  30  \
fund_name              AREA PLAN FOR THE AGING TIT 7   
adopted_budget_amount                              0   

                                                  40  
fund_name              GENERAL FUND (GENERAL BUDGET)  
adopted_budget_amount                         222431

### Values

Generally, we'll be approaching our data in aggregate or filtering by some condition. Sometimes you will want to access an individual value.

My preferred method is to again use `.loc`.

In [29]:
df.loc[0, 'fund_name']

'GENERAL FUND (GENERAL BUDGET)'

There is a more specific method of access, `.at`, but this _only_ allows you to access a single value, and so is less generalizable than `.loc`.

In [30]:
df.at[0, 'fund_name']

'GENERAL FUND (GENERAL BUDGET)'

You may also see the follow format in example code online. We won't go into the details of it now, but essentially we can think of this as selecting a single column AFTER selecting a single row.

In [31]:
df.loc[0]['fund_name']

'GENERAL FUND (GENERAL BUDGET)'

We can also apply this in the opposite order, selecting a single row AFTER selecting a single column.

In [32]:
df['fund_name'].loc[0]

'GENERAL FUND (GENERAL BUDGET)'

#### Pro-tip

Use standard `.loc` notation regardless of whether you are trying to access a DataFrame, Series, or value. It is the most explicit and robust method for selection, and will make your code easy to read and understand.

## Meta-Data Exploration

Now that we understand some of the basics of navigating our data in Pandas, we'll learn some best practices for learning about _what_ our data is.

### Shape

One of the most important things to always check is the `.shape` of our data. This will return the size of our data as a tuple of `(rows, columns)`.

In [33]:
df.shape

(3653, 17)

Here we have 3653 rows and 17 columns.

### Data Types

We can see the data type of each column with the `.dtypes` attribute.

In [34]:
df.dtypes

budget_fiscal_year              int64
department_name                object
fund_name                      object
account_name                   object
adopted_budget_amount         float64
total_expenditures            float64
budget_change_amount          float64
budget_transfer_in_amount     float64
budget_transfer_out_amount    float64
total_budget                  float64
encumbrance_amount            float64
pre_encumbrance_amount        float64
budget_uncommitted_amount     float64
account_group_name             object
fund                           object
account                        object
department                      int64
dtype: object

### Info

Using the `.info` method will give us more detail. This includes:

- A description of our index (here a `RangeIndex` with 3653 entries)
- The total number of columns
- The name, data type, and count of non-null values for each column (this will be truncated in DataFrames with MANY columns).
- The data types present in our DataFrame and the count of columns of each type
- The memory usage of our current DataFrame

Looking at the `.info` of your data is always a great place to start your data exploration.

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3653 entries, 0 to 3652
Data columns (total 17 columns):
budget_fiscal_year            3653 non-null int64
department_name               3653 non-null object
fund_name                     3653 non-null object
account_name                  3651 non-null object
adopted_budget_amount         3653 non-null float64
total_expenditures            2538 non-null float64
budget_change_amount          3653 non-null float64
budget_transfer_in_amount     3653 non-null float64
budget_transfer_out_amount    3653 non-null float64
total_budget                  3653 non-null float64
encumbrance_amount            3653 non-null float64
pre_encumbrance_amount        3653 non-null float64
budget_uncommitted_amount     3653 non-null float64
account_group_name            1375 non-null object
fund                          3653 non-null object
account                       3653 non-null object
department                    3653 non-null int64
dtypes: float64(9),

### Index

Both our rows and columns have an associated index. For our rows, this is the `.index`.

In [36]:
df.index

RangeIndex(start=0, stop=3653, step=1)

Here we have a serial integer range as our index, starting at 0, stopping at 3653 (exclusive), and incrementing by 1 at each row.

### Columns

We access our column index with `.columns`.

In [37]:
df.columns

Index(['budget_fiscal_year', 'department_name', 'fund_name', 'account_name',
       'adopted_budget_amount', 'total_expenditures', 'budget_change_amount',
       'budget_transfer_in_amount', 'budget_transfer_out_amount',
       'total_budget', 'encumbrance_amount', 'pre_encumbrance_amount',
       'budget_uncommitted_amount', 'account_group_name', 'fund', 'account',
       'department'],
      dtype='object')

These are the names of each of the columns in our DataFrame. We can generally treat this object as a list (although sometimes we'll want to explicitly cast it as such for some operations).

To force an index to a list, we can use `.tolist` or just call `list` on it.

In [38]:
df.columns.tolist()

['budget_fiscal_year',
 'department_name',
 'fund_name',
 'account_name',
 'adopted_budget_amount',
 'total_expenditures',
 'budget_change_amount',
 'budget_transfer_in_amount',
 'budget_transfer_out_amount',
 'total_budget',
 'encumbrance_amount',
 'pre_encumbrance_amount',
 'budget_uncommitted_amount',
 'account_group_name',
 'fund',
 'account',
 'department']

In [39]:
list(df.columns)

['budget_fiscal_year',
 'department_name',
 'fund_name',
 'account_name',
 'adopted_budget_amount',
 'total_expenditures',
 'budget_change_amount',
 'budget_transfer_in_amount',
 'budget_transfer_out_amount',
 'total_budget',
 'encumbrance_amount',
 'pre_encumbrance_amount',
 'budget_uncommitted_amount',
 'account_group_name',
 'fund',
 'account',
 'department']

## Methods and Attributes

By now, you're probably wondering:

> "Why do some of these things have `()`, some `[]`, and some no punctuation after the words?"

Great question!

If you've tried to explore Pandas independently, you may have given up in frustration, unable to troubleshoot the errors that you were getting trying to implement the solutions you found on StackOverflow. 

**You're not alone.**

It will take time to familiarize yourself with the syntax for interacting with the Pandas application programming interface (API). In my experience, the biggest difficulty is learning that Pandas is _always trying to do the right thing_. Or rather, it's trying to do the thing it _thinks_ you might want to do.

Pandas was developed to bring the ease of data exploration and manipulation found in Excel, SQL, and R DataFrames into the Python environment. As such, there is a _ton_ of desired functionality that has been implemented by the open source community. Hopefully the following rules can help you begin to organize Pandas syntax in your mind. [(The official Pandas docs are always a good place to inform yourself)](https://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html).

### 1. When Accessing a Static Attribute, No Punctuation Is Needed

A DataFrame is a `Class` object. You may or may not have worked with classes before; that's okay! An important thing to understand about a class is that it can store already computed static values (or attributes) that don't actually transform or operate on the data in anyway at time of access.

Here are some examples we've seen so far:

| Syntax | Attribute accessed |
| --- | --- |
| `.shape` | A tuple of the shape of the data, rows by columns |
| `.index` | The index of the data, as an Index object |
| `.columns` | The columns of the data, as an Index object |
| `.T` | The transpose of the data, with rows as columns and columns as rows |
| `.some_specific_column_name` e.g., `.fund_name` | A column series |

In each of these example, the returned attribute is just some aspect of the defined data that you're accessing. This will make a bit more sense after we consider...

### 2. When Modifying Data or Performing Calculations, Use `()`

Classes also have associated methods. There are the same as functions, but can be thought of to generally have pre-defined arguments that act upon the data in the class. Thus far, most of the methods that we've invoked have _modified a preview of our data_.

Here are some examples we've seen so far:

| Syntax | Method applied |
| --- | --- |
| `.head()` | Limit returned view to first 5 rows |
| `.sample()` | Return a random sample row from our data |
| `.tolist()` | Return data as a list object |
| `.info()` | Return a summary of indices, columns, dtypes, nulls, and size |

Some other methods we'll see shortly extend common operations you may have encountered in SQL or numpy:

| Syntax | Method applied |
| --- | --- |
| `.groupby()` | Group by one or more columns |
| `.count()` | Provide an aggregate count of rows |
| `.mean()` | Calculate the mean of one or more columns |
| `.join()` | Join a DataFrame with another DataFrame on some condition |

Pandas methods should always be used, when possible, as they're designed to be highly optimized for calculations, sorting, and aggregation. We'll learn many of these today, but there are far too many to cover in any one lesson (I still find new, useful methods regularly as complete projects).

**But what about those pesky square brackets?**

### 3. When Filtering or Accessing Data, Use `[]`

This can be a conceptual hurdle for some, but understanding square bracket notation really boils down to two main points:

1. DataFrames are built on top of `numpy` arrays and extend much of the `numpy` notation.
2. DataFrames can be thought of as a dictionary of `Series` and use dictionary indexing notation.

Of course, neither of these concepts is helpful if you aren't familiar with keying into dictionaries and indexing in numpy.

The most important things to keep in mind:

1. Data is accessed by row(s) and then column(s)
    - `df.loc[row]`
    - `df.loc[row, column]`
    - `df.loc[[row1, row2], [column1, column2]]`
2. If no rows are passed, Pandas assumes you are trying to index into your columns
    - `df[column]`
    - `df[[column1, column2]]`
    - `df.some_column_name` (not bracket notation, but the same concept)
3. `:` can be used to select a range
    - `df.loc[:, column]`
    - `df.loc[0:5, column]`
    - `df.loc[0:5, column1:column5]`
4. Double brackets always return a DataFrame.
    - `df[[column]]`
    - `df[[column1, column2]]`
    - `df.loc[[row1, row2], [column1, column2]]`

**This is a lot of information, and you're not expected to remember all of this right now**. One of the nice things about coding in Jupyter is that many of our error messages are informative, and we can often quickly add the punctuation needed to revise our code and get it to run. **The only way to learn is to try and fail**.


In [40]:
df.loc[0:5, 'department_name':'account_name']

department_name                       fund_name  \
0           AGING   GENERAL FUND (GENERAL BUDGET)   
1           AGING   TITLE VII OLDER AMERICANS ACT   
2           AGING   SENIOR HUMAN SERVICES PROGRAM   
3           AGING   AREA PLAN FOR THE AGING TIT 7   
4           AGING   GENERAL FUND (GENERAL BUDGET)   
5           AGING  OMBUDSMAN INITIATIVE PROGRAM F   

                                 account_name  
0                        CONTRACTUAL SERVICES  
1                     OMBUDSMAN VII A PROGRAM  
2                     EVIDENCE BASED PROGRAMS  
3            HOME DELIVERED MEALS FOR SENIORS  
4                            OVERTIME GENERAL  
5  STATE HEALTH FACILITIES CITATION PENALTIES

## Categorical Data

As we've already seen, Pandas allows you to add a number of different kinds of data into the same DataFrame. It has a suite of methods specifically built out for dealing with categorical data.

By default, Pandas will load any column that contains letters as an `object` type.

You can use the `select_dtypes` method to return a view of your DataFrame with only the specified type present.

In [41]:
df.select_dtypes('object').head()

department_name                      fund_name  \
0           AGING  GENERAL FUND (GENERAL BUDGET)   
1           AGING  TITLE VII OLDER AMERICANS ACT   
2           AGING  SENIOR HUMAN SERVICES PROGRAM   
3           AGING  AREA PLAN FOR THE AGING TIT 7   
4           AGING  GENERAL FUND (GENERAL BUDGET)   

                       account_name     account_group_name fund account  
0              CONTRACTUAL SERVICES               EXPENSES  100  003040  
1           OMBUDSMAN VII A PROGRAM                    NaN  564  02PB01  
2           EVIDENCE BASED PROGRAMS                    NaN  42J  02R340  
3  HOME DELIVERED MEALS FOR SENIORS                    NaN  395  02PQ04  
4                  OVERTIME GENERAL  SALARIES AND BENEFITS  100  001090

I like to save the the columns object out so that I can use this to easily operate on all my categorical columns.

Not that when we use a list of column names, the list itself contains a set of square brackets, so we can pass it back to our DataFrame to return a DataFrame.

In [42]:
cat_cols = df.select_dtypes('object').columns
df[cat_cols].head()

department_name                      fund_name  \
0           AGING  GENERAL FUND (GENERAL BUDGET)   
1           AGING  TITLE VII OLDER AMERICANS ACT   
2           AGING  SENIOR HUMAN SERVICES PROGRAM   
3           AGING  AREA PLAN FOR THE AGING TIT 7   
4           AGING  GENERAL FUND (GENERAL BUDGET)   

                       account_name     account_group_name fund account  
0              CONTRACTUAL SERVICES               EXPENSES  100  003040  
1           OMBUDSMAN VII A PROGRAM                    NaN  564  02PB01  
2           EVIDENCE BASED PROGRAMS                    NaN  42J  02R340  
3  HOME DELIVERED MEALS FOR SENIORS                    NaN  395  02PQ04  
4                  OVERTIME GENERAL  SALARIES AND BENEFITS  100  001090

### Summary Statistics

There are limited summary statistics to perform on categorical values. Namely we look at total unique values, total count for each value, and the distribution of our total counts.

The method `.nunique` will gives us the number of unique values in each column.

In [43]:
df[cat_cols].nunique()

department_name         54
fund_name              481
account_name          2057
account_group_name       4
fund                   481
account               2347
dtype: int64

We can then call `.unique` on a specific column to get a set of the values contained.

In [44]:
df['account_group_name'].unique()

array(['EXPENSES', nan, 'SALARIES AND BENEFITS', 'SPECIAL', 'EQUIPMENT'],
      dtype=object)

Note that by default, `.nunique` will ignore null values, while `.unique` will return these.

More often, we will use `.value_counts` to return a count of each unique value in a column.

In [45]:
df['department_name'].value_counts()

NON-DEPARTMENTAL - APPROPRIATIONS TO SPECIAL PURPOSE FUND       680
HOUSING AND COMMUNITY INVESTMENT DEPARTMENT                     288
TRANSPORTATION                                                  267
CULTURAL AFFAIRS                                                229
ECONOMIC AND WORKFORCE DEVELOPMENT DEPARTMENT                   215
CITY CLERK                                                      212
CITY ADMINISTRATIVE OFFICER                                     156
RECREATION AND PARKS - SPECIAL ACCOUNTS                         124
WATER AND POWER                                                 120
NON-DEPARTMENTAL - GENERAL CITY PURPOSES                         99
RECREATION AND PARKS                                             92
MAYOR                                                            88
POLICE                                                           82
NON-DEPARTMENTAL - GENERAL                                       75
NON-DEPARTMENTAL - CAPITAL IMPROVEMENT EXPENSE P

By default, these will be return in descending order.

Note that `.value_counts` also ignores missing values. We can override this behavior (in many Pandas methods) by setting `dropna=False`.

In [46]:
df['account_group_name'].value_counts(dropna=False)

NaN                      2278
SPECIAL                   931
EXPENSES                  287
SALARIES AND BENEFITS     143
EQUIPMENT                  14
Name: account_group_name, dtype: int64

The high number of missing values suggests that this is an optional field.

### String Operations

Pandas has great functionality for applying string methods to text data.

Let's look at our categorical data again.

In [47]:
df[cat_cols].head()

department_name                      fund_name  \
0           AGING  GENERAL FUND (GENERAL BUDGET)   
1           AGING  TITLE VII OLDER AMERICANS ACT   
2           AGING  SENIOR HUMAN SERVICES PROGRAM   
3           AGING  AREA PLAN FOR THE AGING TIT 7   
4           AGING  GENERAL FUND (GENERAL BUDGET)   

                       account_name     account_group_name fund account  
0              CONTRACTUAL SERVICES               EXPENSES  100  003040  
1           OMBUDSMAN VII A PROGRAM                    NaN  564  02PB01  
2           EVIDENCE BASED PROGRAMS                    NaN  42J  02R340  
3  HOME DELIVERED MEALS FOR SENIORS                    NaN  395  02PQ04  
4                  OVERTIME GENERAL  SALARIES AND BENEFITS  100  001090

I don't love that the department names are in all caps. Changing this is as simple as accessing the Series and then invoking `.str.capitalize`. You'll find easy application of your favorite string methods here, alongside a number of Pandas-specific string methods.

This will return a copy of the Series with the changes applied, but nothing will have changed in the original data.

You can chain methods easily in Pandas. Because the returned object is a Series, we can use `.value_counts` to see unique values and counts with this new formatting.

In [48]:
df['department_name'].str.capitalize().value_counts()

Non-departmental - appropriations to special purpose fund       680
Housing and community investment department                     288
Transportation                                                  267
Cultural affairs                                                229
Economic and workforce development department                   215
City clerk                                                      212
City administrative officer                                     156
Recreation and parks - special accounts                         124
Water and power                                                 120
Non-departmental - general city purposes                         99
Recreation and parks                                             92
Mayor                                                            88
Police                                                           82
Non-departmental - general                                       75
Non-departmental - capital improvement expense p

Let's go ahead and save a similar formatting change for the `account_group_name` field.

Just do this by assigning right back into the Series in the DataFrame.

In [49]:
df['account_group_name'] = df['account_group_name'].str.capitalize()

## Numeric Data

There's even greater functionality for numeric data. A great place to start is to look at summary statistics.

### Summary Statistics

Pandas `.describe` method will automatically return most of the desired summary stats for all numeric fields.

In [50]:
df.describe()

budget_fiscal_year  adopted_budget_amount  total_expenditures  \
count              3653.0           3.653000e+03        2.538000e+03   
mean               2018.0           3.887015e+06        1.213486e+07   
std                   0.0           3.781352e+07        1.383864e+08   
min                2018.0           0.000000e+00        0.000000e+00   
25%                2018.0           0.000000e+00        3.423884e+04   
50%                2018.0           0.000000e+00        2.000962e+05   
75%                2018.0           3.000000e+05        1.564424e+06   
max                2018.0           1.114645e+09        5.256445e+09   

       budget_change_amount  budget_transfer_in_amount  \
count          3.653000e+03               3.653000e+03   
mean           1.075858e+06               1.136714e+05   
std            2.627173e+07               1.327327e+06   
min           -5.718038e+07               0.000000e+00   
25%            0.000000e+00               0.000000e+00   
50%            1.065000e+03               0.000000e+00   
75%            1.056430e+05               0.000000e+00   
max            1.449055e+09               3.590700e+07   

       budget_transfer_out_amount  total_budget  encumbrance_amount  \
count                3.653000e+03  3.653000e+03        3.653000e+03   
mean                 1.136714e+05  4.962873e+06        9.763127e+04   
std                  1.878178e+06  4.583206e+07        7.039609e+05   
min                  0.000000e+00  0.000000e+00        0.000000e+00   
25%                  0.000000e+00  6.000000e+03        0.000000e+00   
50%                  0.000000e+00  1.020551e+05        0.000000e+00   
75%                  0.000000e+00  8.000000e+05        0.000000e+00   
max                  8.172030e+07  1.449055e+09        1.602654e+07   

       pre_encumbrance_amount  budget_uncommitted_amount   department  
count            3.653000e+03               3.653000e+03  3653.000000  
mean             6.138717e+03              -3.572344e+06    48.454969  
std              1.000766e+05               1.072887e+08    26.521788  
min              0.000000e+00              -5.256445e+09     2.000000  
25%              0.000000e+00               0.000000e+00    30.000000  
50%              0.000000e+00               4.355400e+02    50.000000  
75%              0.000000e+00               8.000000e+04    62.000000  
max              4.069569e+06               4.289099e+08    98.000000

I find it's often easier to interpret these values once they're transposed.

In [51]:
df.describe().T

count          mean           std           min  \
budget_fiscal_year          3653.0  2.018000e+03  0.000000e+00  2.018000e+03   
adopted_budget_amount       3653.0  3.887015e+06  3.781352e+07  0.000000e+00   
total_expenditures          2538.0  1.213486e+07  1.383864e+08  0.000000e+00   
budget_change_amount        3653.0  1.075858e+06  2.627173e+07 -5.718038e+07   
budget_transfer_in_amount   3653.0  1.136714e+05  1.327327e+06  0.000000e+00   
budget_transfer_out_amount  3653.0  1.136714e+05  1.878178e+06  0.000000e+00   
total_budget                3653.0  4.962873e+06  4.583206e+07  0.000000e+00   
encumbrance_amount          3653.0  9.763127e+04  7.039609e+05  0.000000e+00   
pre_encumbrance_amount      3653.0  6.138717e+03  1.000766e+05  0.000000e+00   
budget_uncommitted_amount   3653.0 -3.572344e+06  1.072887e+08 -5.256445e+09   
department                  3653.0  4.845497e+01  2.652179e+01  2.000000e+00   

                                 25%        50%          75%           max  
budget_fiscal_year           2018.00    2018.00     2018.000  2.018000e+03  
adopted_budget_amount           0.00       0.00   300000.000  1.114645e+09  
total_expenditures          34238.84  200096.24  1564423.615  5.256445e+09  
budget_change_amount            0.00    1065.00   105643.000  1.449055e+09  
budget_transfer_in_amount       0.00       0.00        0.000  3.590700e+07  
budget_transfer_out_amount      0.00       0.00        0.000  8.172030e+07  
total_budget                 6000.00  102055.13   800000.000  1.449055e+09  
encumbrance_amount              0.00       0.00        0.000  1.602654e+07  
pre_encumbrance_amount          0.00       0.00        0.000  4.069569e+06  
budget_uncommitted_amount       0.00     435.54    80000.000  4.289099e+08  
department                     30.00      50.00       62.000  9.800000e+01

Note that two of our numeric fields actually represent categorical data, `budget_fiscal_year` and `department`.

There are a number of ways to deal with this. I prefer to just go ahead and cast them as `object` type. This will omit them from our numeric operations (and also allow us to easily group them with our other categorical columns).

Use `.astype` to change each column to an object and overwrite the original column.

In [52]:
df['department'] = df['department'].astype('object')

In [53]:
df['budget_fiscal_year'] = df['budget_fiscal_year'].astype('object')

Now when we look at our `.describe`, these values won't appear.

In [54]:
df.describe().T

count          mean           std           min  \
adopted_budget_amount       3653.0  3.887015e+06  3.781352e+07  0.000000e+00   
total_expenditures          2538.0  1.213486e+07  1.383864e+08  0.000000e+00   
budget_change_amount        3653.0  1.075858e+06  2.627173e+07 -5.718038e+07   
budget_transfer_in_amount   3653.0  1.136714e+05  1.327327e+06  0.000000e+00   
budget_transfer_out_amount  3653.0  1.136714e+05  1.878178e+06  0.000000e+00   
total_budget                3653.0  4.962873e+06  4.583206e+07  0.000000e+00   
encumbrance_amount          3653.0  9.763127e+04  7.039609e+05  0.000000e+00   
pre_encumbrance_amount      3653.0  6.138717e+03  1.000766e+05  0.000000e+00   
budget_uncommitted_amount   3653.0 -3.572344e+06  1.072887e+08 -5.256445e+09   

                                 25%        50%          75%           max  
adopted_budget_amount           0.00       0.00   300000.000  1.114645e+09  
total_expenditures          34238.84  200096.24  1564423.615  5.256445e+09  
budget_change_amount            0.00    1065.00   105643.000  1.449055e+09  
budget_transfer_in_amount       0.00       0.00        0.000  3.590700e+07  
budget_transfer_out_amount      0.00       0.00        0.000  8.172030e+07  
total_budget                 6000.00  102055.13   800000.000  1.449055e+09  
encumbrance_amount              0.00       0.00        0.000  1.602654e+07  
pre_encumbrance_amount          0.00       0.00        0.000  4.069569e+06  
budget_uncommitted_amount       0.00     435.54    80000.000  4.289099e+08

If you're not familiar, the `e+06` is scientific notation, and just means that the decimal should be moved 6 places to the right.

Here's a table for what each of our calculated values represents:

| statistic | meaning |
| --- | --- |
| count | Number of non-null elements |
| mean | Average of non-null elements |
| std | Standard deviation |
| min | The smallest value in the column |
| 25% | Value greater than 25% of data (lower quartile, Q1) |
| 50% | The middle value (50th percentile, Q2) |
| 75% | Value greater than 75% of data (upper quartile, Q3) |
| max | The largest value in the column |

We note that many of our columns are full of mostly zeros (5 of our columns have a median of 0, and 4 have the value 0 at the 75 percentile).

All of our columns have min values of 0 or less.

The only column with missing values is `total_expenditures`. Because we're looking at 2018 data, it's possible that these values just haven't been calculated and finalized yet, but we should keep this in mind moving forward as we look at other years.

#### Select Numeric Columns

To be explicit moving forward, let's get a quick list of numeric columns, we'll again use `.select_dtypes` with the keyword `exclude='object'`. We'll just directly access our `.columns` from here.

In [55]:
num_cols = df.select_dtypes(exclude='object').columns
num_cols

Index(['adopted_budget_amount', 'total_expenditures', 'budget_change_amount',
       'budget_transfer_in_amount', 'budget_transfer_out_amount',
       'total_budget', 'encumbrance_amount', 'pre_encumbrance_amount',
       'budget_uncommitted_amount'],
      dtype='object')

Here's a convenient table of the description provided alongside the data for each of these fields:

| Column name | Description |
| --- | --- |
| adopted_budget_amount | Original budget amount adopted by Mayor and Council |
| total_expenditures | Total Budget Fiscal Year amount expended from account to date |
| budget_change_amount | Amendment to the adopted budget amount |
| budget_transfer_in_amount | Increase in appropriation to account by transfer in of funds |
| budget_transfer_out_amount | Decrease in appropriation to account by transfer out of funds |
| total_budget | Appropriation account amount net of changes and transfers to/from the original budgeted amount |
| encumbrance_amount | Obligation or commitment to pay for a good or service |
| pre_encumbrance_amount | Anticipated obligation or commitment to pay for a good or service |
| budget_uncommitted_amount | Total unused appropriation after expenditures and encumbrances |

### Aggregate Statistics

We can also calculate these same statistics on each column, or the entire DataFrame. We'll run through these operations quickly as a demonstration.

#### `.count`

`.count` will work on both numeric and categorical values. Note that nulls are ignored.

In [56]:
df.count()

budget_fiscal_year            3653
department_name               3653
fund_name                     3653
account_name                  3651
adopted_budget_amount         3653
total_expenditures            2538
budget_change_amount          3653
budget_transfer_in_amount     3653
budget_transfer_out_amount    3653
total_budget                  3653
encumbrance_amount            3653
pre_encumbrance_amount        3653
budget_uncommitted_amount     3653
account_group_name            1375
fund                          3653
account                       3653
department                    3653
dtype: int64

#### `.mean`

`.mean` will only evaluated numeric values, ignoring nulls.

In [57]:
df[num_cols].mean()

adopted_budget_amount         3.887015e+06
total_expenditures            1.213486e+07
budget_change_amount          1.075858e+06
budget_transfer_in_amount     1.136714e+05
budget_transfer_out_amount    1.136714e+05
total_budget                  4.962873e+06
encumbrance_amount            9.763127e+04
pre_encumbrance_amount        6.138717e+03
budget_uncommitted_amount    -3.572344e+06
dtype: float64

By default, means are calculated for each column. However, we can changed this to be calculated over rows by passing an `axis=1` keyword argument.

In [58]:
df[num_cols].mean(axis=1).head()

0    693405.111111
1     29292.000000
2    101149.000000
3    847948.333333
4      5733.333333
dtype: float64

In this investigation, these numbers aren't extremely informative. As a reminder, let's look at our numeric columns again.

In [59]:
num_cols

Index(['adopted_budget_amount', 'total_expenditures', 'budget_change_amount',
       'budget_transfer_in_amount', 'budget_transfer_out_amount',
       'total_budget', 'encumbrance_amount', 'pre_encumbrance_amount',
       'budget_uncommitted_amount'],
      dtype='object')

So the row-wise means we are returning are just being calculated on these various numbers. Again, this is for demonstration, and not really informative.

#### `.std`

Standard deviation will only be calculated on numeric columns.

In [60]:
df[num_cols].std()

adopted_budget_amount         3.781352e+07
total_expenditures            1.383864e+08
budget_change_amount          2.627173e+07
budget_transfer_in_amount     1.327327e+06
budget_transfer_out_amount    1.878178e+06
total_budget                  4.583206e+07
encumbrance_amount            7.039609e+05
pre_encumbrance_amount        1.000766e+05
budget_uncommitted_amount     1.072887e+08
dtype: float64

#### Quantiles

The `.quantile` method will allow you to specify any value between 0 and 1.

In [61]:
df[num_cols].quantile(.3)

adopted_budget_amount             0.0
total_expenditures            42000.0
budget_change_amount              0.0
budget_transfer_in_amount         0.0
budget_transfer_out_amount        0.0
total_budget                  13409.6
encumbrance_amount                0.0
pre_encumbrance_amount            0.0
budget_uncommitted_amount         0.0
Name: 0.3, dtype: float64

We also have the built-in `.median` function for the 50 percentile.

In [62]:
df[num_cols].median()

adopted_budget_amount              0.00
total_expenditures            200096.24
budget_change_amount            1065.00
budget_transfer_in_amount          0.00
budget_transfer_out_amount         0.00
total_budget                  102055.13
encumbrance_amount                 0.00
pre_encumbrance_amount             0.00
budget_uncommitted_amount        435.54
dtype: float64

#### `.min` and `.max`

These will work on both numeric and categorical columns, returning minimum and maximum values for each column. For categorical columns, these are alphabetically sorted.

In [63]:
df.min()

budget_fiscal_year                                                  2018
department_name                                                    AGING
fund_name                     100 RESILIENT CITIES INITIATIVE GRANT FUND
adopted_budget_amount                                                  0
total_expenditures                                                     0
budget_change_amount                                        -5.71804e+07
budget_transfer_in_amount                                              0
budget_transfer_out_amount                                             0
total_budget                                                           0
encumbrance_amount                                                     0
pre_encumbrance_amount                                                 0
budget_uncommitted_amount                                   -5.25645e+09
fund                                                                 100
account                                            

In [64]:
df.max()

budget_fiscal_year                                 2018
department_name                                     ZOO
fund_name                     ZOO ENTERPRISE TRUST FUND
adopted_budget_amount                       1.11464e+09
total_expenditures                          5.25645e+09
budget_change_amount                        1.44906e+09
budget_transfer_in_amount                    3.5907e+07
budget_transfer_out_amount                  8.17203e+07
total_budget                                1.44906e+09
encumbrance_amount                          1.60265e+07
pre_encumbrance_amount                      4.06957e+06
budget_uncommitted_amount                    4.2891e+08
fund                                                W88
account                                          988210
department                                           98
dtype: object

Note that all of these methods can be applied on a Series as well.

In [65]:
df.total_budget.max()

1449055000.0

### Basic Math Operations

Pandas is set up to do vectorized math operations by default.

We'll go through a few of these here.

#### Addition

In [66]:
(df['budget_change_amount'] + df['adopted_budget_amount']).head()

0    2231382.0
1      87876.0
2     303447.0
3    2543845.0
4       3900.0
dtype: float64

#### Subtraction

In [67]:
(df.total_budget - df.total_expenditures).head()

0    169724.96
1         0.00
2     11109.00
3    124683.00
4      1256.64
dtype: float64

#### Multiplication

In [68]:
(df['budget_transfer_in_amount'] * df['budget_transfer_out_amount']).head()

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
dtype: float64

#### Division
We can also do scalar operations:

In [69]:
(df.budget_change_amount / 12).head()

0       750.000000
1      7323.000000
2     25287.250000
3    211987.083333
4         0.000000
Name: budget_change_amount, dtype: float64

## Sorting

Sorting in Pandas is accomplished easily by using the `sort_values` method. By default, all columns in the DataFrame will be returned with rows sorted in ascending order by the specified column.

Here, let's sort on `total_budget`.

In [70]:
df.sort_values('total_budget').head()

budget_fiscal_year                                    department_name  \
3575               2018                                    WATER AND POWER   
1666               2018                                              MAYOR   
622                2018                                      CITY PLANNING   
2508               2018  NON-DEPARTMENTAL - CAPITAL IMPROVEMENT EXPENSE...   
3572               2018                                    WATER AND POWER   

                                              fund_name  \
3575                     WS RB 2009 SERIES A BOND SERVC   
1666  2013 SECOND CHANCE ACT JUVENILE REENTRY GRANT ...   
622                      PLANNING CASE PROCESSING SPREV   
2508                      GENERAL FUND (GENERAL BUDGET)   
3572                     PWR SYST RB 2012 C BD SER FUND   

                 account_name  adopted_budget_amount  total_expenditures  \
3575  RETIREMENT OF PRINCIPAL                    0.0           3750000.0   
1666                    MAYOR                    0.0                 NaN   
622       BUILDING AND SAFETY               200000.0                 0.0   
2508       CD 8 COST RECOVERY                    0.0                 NaN   
3572         INTEREST EXPENSE                    0.0                 NaN   

      budget_change_amount  budget_transfer_in_amount  \
3575                   0.0                        0.0   
1666                   0.0                        0.0   
622              -200000.0                        0.0   
2508                   0.0                        0.0   
3572                   0.0                        0.0   

      budget_transfer_out_amount  total_budget  encumbrance_amount  \
3575                         0.0           0.0                 0.0   
1666                         0.0           0.0                 0.0   
622                          0.0           0.0                 0.0   
2508                         0.0           0.0                 0.0   
3572                         0.0           0.0                 0.0   

      pre_encumbrance_amount  budget_uncommitted_amount account_group_name  \
3575                     0.0                 -3750000.0                NaN   
1666                     0.0                        0.0                NaN   
622                      0.0                        0.0                NaN   
2508                     0.0                        0.0            Special   
3572                     0.0                        0.0                NaN   

     fund account department  
3575  M9B  988110         98  
1666  56H  46P146         46  
622   52D  68108P         68  
2508  100  00P297         54  
3572  J6B  988210         98

When we sort on a text column, this will be sorted alphabetically.

Let's trying this with department name.

In [71]:
df.sort_values('department_name').head()

budget_fiscal_year department_name                      fund_name  \
0                2018           AGING  GENERAL FUND (GENERAL BUDGET)   
23               2018           AGING   HEALTH INS COUNS ADV (HICAP)   
24               2018           AGING  AREA PLAN FOR THE AGING TIT 7   
25               2018           AGING  SENIOR HUMAN SERVICES PROGRAM   
26               2018           AGING  AREA PLAN FOR THE AGING TIT 7   

                        account_name  adopted_budget_amount  \
0               CONTRACTUAL SERVICES              2222382.0   
23                    HICAP 12 MONTH                    0.0   
24     PREVENTIVE HEALTH FOR SENIORS                    0.0   
25                             AGING                    0.0   
26  NSIP HOME DELIVERED MEALS III C2                    0.0   

    total_expenditures  budget_change_amount  budget_transfer_in_amount  \
0           1608157.04                9000.0                        0.0   
23           362612.00              402384.0                        0.0   
24            15720.00               15720.0                        0.0   
25            72974.84              129399.0                        0.0   
26           532420.00              537683.0                        0.0   

    budget_transfer_out_amount  total_budget  encumbrance_amount  \
0                     453500.0     1777882.0             93331.0   
23                         0.0      402384.0                 0.0   
24                         0.0       15720.0                 0.0   
25                         0.0      129399.0                 0.0   
26                         0.0      537683.0              5263.0   

    pre_encumbrance_amount  budget_uncommitted_amount account_group_name fund  \
0                      0.0                   76393.96           Expenses  100   
23                     0.0                   39772.00                NaN  47Y   
24                     0.0                       0.00                NaN  395   
25                     0.0                   56424.16                NaN  42J   
26                     0.0                       0.00                NaN  395   

   account department  
0   003040          2  
23  02PD12          2  
24  02PQ06          2  
25  02P102          2  
26  02PA08          2

To reverse the sort order, we pass the argument `ascending=False`.

In [72]:
df.sort_values('department_name', ascending=False).head()

budget_fiscal_year department_name                      fund_name  \
3652               2018             ZOO  GENERAL FUND (GENERAL BUDGET)   
3640               2018             ZOO      ZOO ENTERPRISE TRUST FUND   
3628               2018             ZOO      ZOO ENTERPRISE TRUST FUND   
3629               2018             ZOO  GENERAL FUND (GENERAL BUDGET)   
3630               2018             ZOO      ZOO ENTERPRISE TRUST FUND   

                                    account_name  adopted_budget_amount  \
3652               VETERINARY SUPPLIES & EXPENSE               400000.0   
3640               INFORMATION TECHNOLOGY AGENCY                    0.0   
3628                    ANIMAL PURCHASES & SALES                    0.0   
3629  MAINTENANCE MATERIALS, SUPPLIES & SERVICES               821999.0   
3630                      GLAZA MARKETING REFUND              1281682.0   

      total_expenditures  budget_change_amount  budget_transfer_in_amount  \
3652           309926.01                   0.0                        0.0   
3640            21311.00               21311.0                        0.0   
3628                 NaN                   0.0                        0.0   
3629           758065.23                   0.0                        0.0   
3630          1281682.00                   0.0                        0.0   

      budget_transfer_out_amount  total_budget  encumbrance_amount  \
3652                         0.0      400000.0            37915.63   
3640                         0.0       21311.0                0.00   
3628                         0.0           0.0                0.00   
3629                         0.0      821999.0            15634.22   
3630                         0.0     1281682.0                0.00   

      pre_encumbrance_amount  budget_uncommitted_amount account_group_name  \
3652                     0.0                   52158.36           Expenses   
3640                     0.0                       0.00                NaN   
3628                     0.0                       0.00                NaN   
3629                     0.0                   48299.55           Expenses   
3630                     0.0                       0.00            Special   

     fund account department  
3652  100  004570         87  
3640  40E  87P132         87  
3628  40E  879260         87  
3629  100  003160         87  
3630  40E  87P300         87

Since this returns a DataFrame, we could choose to preview only the column that we're interested in, even though we're sorting on a different column.

Here, let's return the `fund_name` for 5 rows with the higheset `adopted_budget_amount`.

In [73]:
df.sort_values('adopted_budget_amount', ascending=False)['fund_name'].head()

2564    GENERAL FUND (GENERAL BUDGET)
2864    GENERAL FUND (GENERAL BUDGET)
586         CITY EMPLOYEES RETIREMENT
283          TX REV ANTICIP NOTE PROC
285          TX REV ANTICIP NOTE PROC
Name: fund_name, dtype: object

Of course, it might be more useful to also know what the `adopted_budget_amount` is here. Let's return just these two columns, again sorted to show the 5 largest amounts.

In [74]:
df.sort_values('adopted_budget_amount', ascending=False)[['fund_name', 'adopted_budget_amount']].head()

fund_name  adopted_budget_amount
2564  GENERAL FUND (GENERAL BUDGET)           1.114645e+09
2864  GENERAL FUND (GENERAL BUDGET)           1.108217e+09
586       CITY EMPLOYEES RETIREMENT           1.007502e+09
283        TX REV ANTICIP NOTE PROC           6.485940e+08
285        TX REV ANTICIP NOTE PROC           4.599837e+08

## Lesson Summary

After today's lesson you should feel comfortable with the following basics of Pandas:

1. Module import
1. Data import
1. Previewing Data
1. Differences between DataFrames and Series
1. Selecting rows and columns
1. Basic metadata exploration
1. Differences between methods and attributes
1. Basic methods for categorical data
1. Basic methods for numeric data
1. Sorting

Please proceed to complete the included lab to test these skills.